pkg/controller/common.go

```diff
@@ -68,17 +79,314 @@ type CloudProvisioner interface {
+		}
+	}
+	sort.Slice(nodes[:], func(i, j int) bool {
+		return nodeScores[nodes[i]] > nodeScores[nodes[j]]
+	})
+
+	if len(nodes) < numReplica {
+		selected, err := selectNodes(ctx, azr, numReplica-len(nodes), primaryNode, nodeScores)
+		if err != nil {
+			return nil, err
+		}
+		nodes = append(nodes, selected...)
+	}
+	return nodes[:min(len(nodes), numReplica)], nil
+}
+
+func (c *SharedState) addPod(pod *corev1.Pod) {
+	podKey := getQualifiedName(pod.Namespace, pod.Name)
+	claims := []string{}
+	for _, volume := range pod.Spec.Volumes {
+		if volume.CSI == nil || volume.CSI.Driver != azureutils.DriverName {
+			continue
+		}
+		namespacedClaimName := getQualifiedName(pod.Namespace, volume.PersistentVolumeClaim.ClaimName)
+		claims = append(claims, namespacedClaimName)
+		v, _ := c.claimToPodsMap.LoadOrStore(namespacedClaimName, []string{})
+
+		pods := v.([]string)
+		podExist := false
+		for _, pod := range pods {
+			if pod == podKey {
+				podExist = true
+				break
+			}
+		}
+		if !podExist {
+			pods = append(pods, podKey)
+		}
+		c.claimToPodsMap.Store(namespacedClaimName, pods)
+	}
+	c.podToClaimsMap.Store(podKey, claims)
+}
+
+func (c *SharedState) deletePod(podKey string) {
+	value, exists := c.podToClaimsMap.LoadAndDelete(podKey)
+	if !exists {
+		return
+	}
+	claims := value.([]string)
+
+	for _, claim := range claims {
+		c.claimToPodsMap.Delete(claim)
+	}
+}
+
+func (c *SharedState) addVolumeAndClaim(azVolumeName, pvClaimName string) {
+	c.volumeToClaimMap.Store(azVolumeName, pvClaimName)
+	c.claimToVolumeMap.Store(pvClaimName, azVolumeName)
+}
+
+func selectNodes(ctx context.Context, azr azReconciler, numNodes int, primaryNode string, exceptionNodes map[string]int) ([]string, error) {
+	// TODO: fill with node selection logic (based on node capacity and node failure domain)
+	/*
+		Below is temporary node selection logic
+	*/
+	filteredNodes := []string{}
+	var nodes *v1alpha1.AzDriverNodeList
+	var err error
+	// List all AzDriverNodes
+	nodes = &v1alpha1.AzDriverNodeList{}
+	if err = azr.getClient().List(ctx, nodes, &client.ListOptions{}); err != nil {
+		klog.Errorf("failed to retrieve azDriverNode List for namespace %s: %v", azr.getNamespace(), err)
+		return filteredNodes, err
+	}
+
+	nodeScores := map[string]int{}
+	if nodes != nil {
+		for _, node := range nodes.Items {
+			if _, ok := exceptionNodes[node.Name]; ok || node.Name == primaryNode {
+				continue
+			}
+
+			// filter out attachments labeled with specified node and volume
+			var attachmentList v1alpha1.AzVolumeAttachmentList
+			nodeRequirement, err := createLabelRequirements(azureutils.NodeNameLabel, node.Name)
+			if err != nil {
+				klog.Errorf("Encountered error while creating Requirement: %+v", err)
+				continue
+			}
+
+			labelSelector := labels.NewSelector().Add(*nodeRequirement)
+			if err := azr.getClient().List(ctx, &attachmentList, &client.ListOptions{LabelSelector: labelSelector}); err != nil {
```

Comment:
If there are no attachments for that node, will it return an error?

Reply:
No, it will just be an empty list